extensions-core/azure-extensions/src/main/java/org/apache/druid/storage/azure/AzureStorage.java

```diff
@@ -56,255 +54,391 @@
  */
 public class AzureStorage
```

Comment:
is it possible for a cluster using azure as deep storage to first make a read-only call to azure (therefore not creating the container) and for the function to throw a exception instead of creating the container and returning a empty list or something?

like e.g if i start a new cluster in azure with no container what's the first api call that gets made and what happens if its a read-only call? i think i would be more concerned with error handling, if we return a good exception that says "the container doesn't exist" or something like this, then the operator could just create the container at that point.

Reply:
to be specific, what does getBlobContainerClient do if the container doesn't exist?

Reply:
thinking about it a bit more, the first write operation is always triggered by a task, either a msq task pushing intermediate segments, or the middleManager/overlord pushing task logs.

this seems ok to me, the only class of things i can think of failing are like, e.g. the task logs cleaner. but if that throws a exception it would be fine since no tasks would have been run.

Reply:
That's a valid question.

- `getBlobContainerClient` does nothing yet. It simply creates a client that will be used for an API call later.
- `createBlobContainerIfNotExists` first [calls](https://github.com/Azure/azure-sdk-for-java/blob/5648561dc792a993ec0fb4b041d76dc64971b9bb/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobServiceClient.java#L206-L211) `getBlobContainerClient`, and then right away executes an API call (`createIfNotExistsWithResponse`) for that client.

In the current implementation, if you _configure_ a container that does not exist, the calls that I found call `azureStorage.getBlockBlobExists` first, which then simply returns `false` (called for streaming task files, or from the storage layer).